src/swapd/swap_state.rs

```diff
@@ -1854,80 +1794,86 @@ fn try_alice_canceled_to_alice_refund_or_alice_punish(
-                tx.clone(),
-                runtime.swap_id,
-                runtime.monero_address_creation_height,
-            )?;
-            // Check if we already registered the lock transaction, if so, initiate sweeping procedure
-            runtime.log_debug(format!("{:?}", runtime.syncer_state.confirmations));
-            if runtime
-                .syncer_state
-                .confirmations
-                .get(&TxLabel::AccLock)
-                .is_some()
-            {
-                let task = runtime.syncer_state.sweep_xmr(sweep_xmr.clone(), true);
-                let sweep_address = if let Task::SweepAddress(sweep_address) = task {
-                    sweep_address
-                } else {
-                    return Ok(None);
-                };
-                runtime.log_monero_maturity(sweep_xmr.destination_address);
-                runtime.log_warn(
-                    "Peerd might crash, just ignore it, counterparty closed \
-                        connection but you don't need it anymore!",
-                );
-                Ok(Some(SwapStateMachine::AliceRefund(sweep_address)))
-            } else {
-                if runtime.syncer_state.awaiting_funding {
-                    runtime.log_warn(
-                        "FundingCompleted never emitted, emitting it now to clean up farcasterd",
-                    );
-                    runtime.syncer_state.awaiting_funding = false;
-                    event.send_ctl_service(
-                        ServiceId::Farcasterd,
-                        CtlMsg::FundingCompleted(Blockchain::Monero),
+                // When Alice learns of the refund transaction, immediately extract the Monero keys from its adaptor signature
+                Some(&TxLabel::Refund)
+                    if runtime
+                        .syncer_state
+                        .tasks
+                        .txids
+                        .contains_key(&TxLabel::Refund) =>
+                {
+                    runtime.log_debug("Subscribe Refund address task");
+                    let tx = bitcoin::Transaction::deserialize(
+                        &tx.iter().flatten().copied().collect::<Vec<u8>>(),
                     )?;
+
+                    log_tx_seen(runtime.swap_id, &TxLabel::Refund, &tx.txid().into());
+                    let sweep_xmr = wallet.process_refund_tx(
+                        event.endpoints,
+                        tx,
+                        runtime.swap_id,
+                        runtime.monero_address_creation_height,
+                    )?;
+
+                    runtime
+                        .syncer_state
+                        .tasks
+                        .txids
+                        .remove_entry(&TxLabel::Refund);
+
+                    // Check if we already registered the lock transaction, if so, initiate sweeping procedure
+                    runtime.log_debug(format!("{:?}", runtime.syncer_state.confirmations));
```

Comment:
This part didn't change but is it necessary to have `{:?}` fmt?

Reply:
It's a HashMap, not sure how else to handle this.